src/librustc/mir/tcx.rs

```diff
@@ -205,14 +205,19 @@ impl<'tcx> NeoPlace<'tcx> {
         mir: &'cx Mir<'tcx>,
         tcx: &TyCtxt<'cx, 'gcx, 'tcx>,
     ) -> Option<Field> {
-        let (place, by_ref) = if let Some(ProjectionElem::Deref) = self.elems.last() {
-            (self.projection_base(*tcx), true)
-        } else {
-            (self.clone(), false)
-        };
+        let mut by_ref = false;
+        let mut elems = self.elems.iter().rev();
+
+        if let Some(ProjectionElem::Deref) = elems.next() {
+            by_ref = true;
+            elems.next();
+        }
 
-        if let Some(ProjectionElem::Field(field, _ty)) = place.elems.last() {
-            let base_ty = place.projection_base(*tcx).ty(mir, *tcx).to_ty(*tcx);
+        if let Some(ProjectionElem::Field(field, _ty)) = elems.next() {
+            let mut base_ty = self.base.ty(mir);
+            for elem in elems.rev() {
+                base_ty = PlaceTy::from(base_ty).projection_ty(*tcx, elem).to_ty(*tcx);
+            }
```

Comment:
we should at minimum make a helper for this loop

Reply:
something like

`self.base.ty_with_projections(mir, elems.rev())`

Reply:
(But I do agree that the iterator based version of the code is pretty decent and we don't need the "tree helpers" here that I was talking about)